src/watchdog.rs

```diff
@@ -0,0 +1,210 @@
+        ) -> Self {
+            Self {
+                duration,
+                panic_on_trigger,
+                #[cfg(not(esp_idf_version_major = "4"))]
+                subscribed_idle_tasks,
+            }
+        }
+
+        #[cfg(esp_idf_version_major = "4")]
+        pub const fn new(duration: core::time::Duration, panic_on_trigger: bool) -> Self {
+            Self {
+                duration,
+                panic_on_trigger,
+            }
+        }
+    }
+
+    impl Default for WatchdogConfig {
+        fn default() -> Self {
+            Self {
+                duration: core::time::Duration::from_secs(CONFIG_ESP_TASK_WDT_TIMEOUT_S as u64),
+                panic_on_trigger: cfg!(esp_idf_esp_task_wdt_panic), // CONFIG_ESP_TASK_WDT_PANIC
+                #[cfg(not(esp_idf_version_major = "4"))]
+                subscribed_idle_tasks: {
+                    let subscribed_idle_tasks = heapless::Vec::new();
+                    if cfg!(esp_idf_esp_task_wdt_check_idle_task_cpu0) {
+                        subscribed_idle_tasks
+                            .push(crate::task::IdleTask::Core0)
+                            .unwrap();
+                    }
+                    #[cfg(any(esp32, esp32s3))]
+                    if cfg!(esp_idf_esp_task_wdt_check_idle_task_cpu1) {
+                        subscribed_idle_tasks
+                            .push(crate::task::IdleTask::Core1)
+                            .unwrap();
+                    }
+                    subscribed_idle_tasks
+                },
+            }
+        }
+    }
+
+    #[cfg(not(esp_idf_version_major = "4"))]
+    impl From<WatchdogConfig> for esp_task_wdt_config_t {
+        fn from(value: WatchdogConfig) -> Self {
+            esp_task_wdt_config_t {
+                timeout_ms: value.duration.as_millis() as u32,
+                trigger_panic: value.panic_on_trigger,
+                idle_core_mask: value
+                    .subscribed_idle_tasks
+                    .iter()
+                    .fold(0u32, |mask, core| mask & (1 << (*core as u32))),
+            }
+        }
+    }
+
+    pub trait Twdt {}
+
+    pub struct TWDTDriver<'d>(PhantomData<&'d mut ()>);
+
+    unsafe impl Send for TWDTDriver<'_> {}
+
+    static TWDT_DRIVER_REF_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+    impl Clone for TWDTDriver<'_> {
+        fn clone(&self) -> Self {
+            TWDT_DRIVER_REF_COUNT.fetch_add(1, core::sync::atomic::Ordering::Relaxed);
+            Self(Default::default())
+        }
+    }
+
+    impl<'d> TWDTDriver<'d> {
+        pub fn new(
+            _twdt: impl Peripheral<P = TWDT> + 'd,
+            config: &WatchdogConfig,
+        ) -> Result<Self, EspError> {
+            init_or_reconfigure(config)?;
+            Ok(Self(Default::default()))
+        }
+
+        // will fail in isr
+        pub fn watch_current_task(&mut self) -> Result<WatchdogSubscription<'_>, EspError> {
```

Comment:
* The fact that it will fail in ISR is no big deal. Most methods of `esp-idf-hal` and ESP IDF do fail when called from an ISR. Using ISRs is totally unsafe anyway.